prevent mixup of account timelines

connyduck opened the pull request:
This does 2 things:

- Removes `AccountSwitchInterceptor`, the main culprit for the bug. APIs can no longer change their base url after they have been created. As a result they are not Singletons anymore.
- Additionally, I refactored how MainActivity handles Intents to make it less likely to have multiple instances of it active.

Here is how I could reliably reproduce the bug:

- Be logged in with account A and B
- Write a post with account A, cancel it before it sends (go into flight mode for that)
- Switch to account B
- Open the "this post failed to send" notification from account A, drafts will open
- Go back. You are in the MainActivity of account A, everything seems fine.
- Go back again. You are in the old, now broken MainActivity of account B. It uses the database of account B but the network of account A. Refreshing will show posts from A.

closes #4567 
closes #4554
closes #4402 
closes #4148
closes #2663
and possibly #4588

connyduck commented on app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt:
```diff
@@ -66,6 +68,18 @@ object NetworkModule {
 
     private const val TAG = "NetworkModule"
 
+    @Provides
+    @Named("defaultPort")
+    fun providesDefaultPort(): Int {
+        return 443
+    }
+
+    @Provides
+    @Named("defaultScheme")
+    fun providesDefaultScheme(): String {
+        return "https://"
+    }
+
```
This is kinda weird, but the only way I found to have tests run on a random port without https.
Better ideas very welcome!

charlag replied:
could we provide the whole URL depending on the current account?

Tak commented on app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt:
```diff
@@ -125,43 +138,100 @@ object NetworkModule {
+                }
+
+                print(originalRequest)
+                print(request)
```
Intended?

connyduck replied:
nope, thx

charlag commented on app/src/main/java/com/keylesspalace/tusky/MainActivity.kt:
```diff
@@ -363,6 +363,11 @@ class MainActivity : BottomSheetActivity(), ActionButtonActivity, MenuProvider {
 
+    override fun onDestroy() {
+        super.onDestroy()
+        cacheUpdater.stop()
```
I think it should be stopped first and then the rest but idk

charlag commented on app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt:
```diff
@@ -125,43 +138,97 @@ object NetworkModule {
+                }
+            }
+            .callFactory { originalRequest ->
+                var request = originalRequest
```
i think that's a lot of code for DI module tbh, it shows in test that you ask it from DI which seems bit off to me, but not critical ofc, we can refactor this later

connyduck commented:
Ok, I perf tested this and found no problems. There are a few KBs more memory used, activity start time is the same as before.

I fixed a bug where sometimes the wrong api instance was used because the containing object was a Singleton and the api was not recreated after account switching.

And I refactored a bit to have less code in the NetworkModule and make the testing simpler.

Please have another look.
